info-compiler/src/main/java/com/google/sps/webcrawler/WebCrawler.java

```diff
@@ -155,80 +160,116 @@ public void compileNewsArticle(String candidateName, String candidateId) throws
+    for (String potentialDateMetatag : CUSTOM_SEARCH_PUBLISHED_DATE_METATAGS) {
+      if (!metadata.has(potentialDateMetatag)) {
+        continue;
+      }
+      String date = metadata.get(potentialDateMetatag).getAsString();
+      for (DateTimeFormatter potentialDateTimeFormatter : DATE_TIME_FORMATTERS) {
+        try {
+          return Date.from(Instant.from(potentialDateTimeFormatter.parse(date)));
+        } catch (IllegalArgumentException | DateTimeException e) {}
+      }
     }
-    return urls;
+    return null;
   }
 
   /**
    * Checks robots.txt for permission to web-scrape, scrapes webpage if permitted and extracts
-   * textual content. Returns an empty {@code Optional<NewsArticle>} in the event of an exception.
+   * textual content to put into {@code newsArticle}. Sets "content" to empty in the event of an
+   * exception.
    */
-  public Optional<NewsArticle> scrapeAndExtractFromHtml(URL url) {
+  public void scrapeAndExtractFromHtml(NewsArticle newsArticle) {
     try {
+      URL url = new URL(newsArticle.getUrl());
       URL robotsUrl = new URL(url.getProtocol(), url.getHost(), "/robots.txt");
       InputStream robotsTxtStream = setTimeoutAndOpenStream(robotsUrl);
       RobotsTxt robotsTxt = RobotsTxt.read(robotsTxtStream);
       robotsTxtStream.close();
       String webpagePath = url.getPath();
       Grant grant = robotsTxt.ask("*", webpagePath);
-      return politelyScrapeAndExtractFromHtml(grant, robotsUrl, url);
+      politelyScrapeAndExtractFromHtml(grant, robotsUrl, newsArticle);
     } catch (Exception e) {
       System.out.println("[ERROR] Error occured in scrapeAndExtractHtml(): " + e);
-      return Optional.empty();
+      newsArticle.setContent("");
     }
   }
 
   /**
    * Checks robots.txt for permission to web-scrape, scrapes webpage if permitted and extracts
-   * textual content. Returns an empty {@code Optional<NewsArticle>} in the event of an exception.
+   * textual content to put into {@code newsArticle}. Sets "content" to empty in the event of an
+   * exception.
    */
-  Optional<NewsArticle> politelyScrapeAndExtractFromHtml(Grant grant, URL robotsUrl, URL url) {
+  void politelyScrapeAndExtractFromHtml(Grant grant, URL robotsUrl,
+      NewsArticle newsArticle) {
     try {
       // Check permission to access and respect the required crawl delay.
-      if (grant == null || grant.hasAccess()) {
-        if (grant != null
-            && grant.getCrawlDelay() != null
-            && !waitForAndSetCrawlDelay(grant, robotsUrl.toString())) {
-          return Optional.empty();
-        }
-        InputStream webpageStream = setTimeoutAndOpenStream(url);
-        Optional<NewsArticle> potentialNewsArticle =
-            newsContentExtractor.extractContentFromHtml(webpageStream, url.toString());
+      if (grant == null
+          || (grant.hasAccess()
+              && (grant.getCrawlDelay() == null
+                  || waitForAndSetCrawlDelay(grant, robotsUrl.toString())))) {
```

Comment:
Does it make sense to move the grant.getCrawlDelay() == null check to waitForAndSetCrawlDelay? that way you are checking less conditions here

Reply:
Will do!